internal/util/logging/attrs.go

```diff
@@ -0,0 +1,97 @@
+// Copyright 2021 FerretDB Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package logging
+
+import (
+	"log/slog"
+	"slices"
+
+	"github.com/FerretDB/FerretDB/v2/internal/util/lazyerrors"
+)
+
+// groupOrAttrs contains group name or attributes.
+type groupOrAttrs struct {
+	group string
+	attrs []slog.Attr
+}
+
+// attrs returns record attributes, as well as handler attributes from goas in map.
+// Attributes with duplicate keys are overwritten, and the order of keys is ignored.
+//
+// If goas contains `name` attribute, the last one in the slice is validated to be string,
+// and returned to be used as logger name. The returned name is not present in returned map.
+// TODO https://github.com/FerretDB/FerretDB/issues/4431
+func attrs(r slog.Record, goas []groupOrAttrs) (map[string]any, string, error) {
+	m := make(map[string]any, r.NumAttrs())
+
+	r.Attrs(func(attr slog.Attr) bool {
+		if attr.Key != "" {
+			m[attr.Key] = resolve(attr.Value)
+
+			return true
+		}
+
+		if attr.Value.Kind() == slog.KindGroup {
+			for _, gAttr := range attr.Value.Group() {
+				m[gAttr.Key] = resolve(gAttr.Value)
+			}
+		}
+
+		return true
+	})
+
+	var name string
+
+	for _, goa := range slices.Backward(goas) {
+		if goa.group != "" && len(m) > 0 {
+			m = map[string]any{goa.group: m}
+			continue
+		}
+
+		for _, attr := range goa.attrs {
+			if attr.Key != nameKey || name != "" {
+				m[attr.Key] = resolve(attr.Value)
+				continue
+			}
+
+			v, ok := attr.Value.Any().(string)
+			if !ok {
+				return nil, "", lazyerrors.Errorf("attribute %q should be a string but was %T", nameKey, v)
```

Comment:
I think logger should not return an error lightly. Often logging is the last result for debugging something not working, in that case, any written log would be useful :hugs:.

For this, how about using `Printf()` or something to make the value into string? Then we can keep the function signature without an error.

Reply:
I think we want to omit situation where values that're not strings are just `Sprintf`d into string. That'd result in really weird logger names if for example name is a struct, or pointer.

To eliminate error from the function, I'd suggest to treat `name` as any other attribute in cases where it's not a string. PTAL at the latest PR

Reply:
Much better! Thank you!